Mathlib/AlgebraicGeometry/EllipticCurve/Modular/TateNormalForm.lean

```diff
@@ -0,0 +1,268 @@
+/-
+Copyright (c) 2025 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import Mathlib.AlgebraicGeometry.EllipticCurve.Affine.Point
+
+
+/-!
+
+# Tate normal form of elliptic curves
+
+This file defines the Tate normal form of Weierstrass equations of elliptic curves. It
+parametrises elliptic curves with a given point `P` with `P, 2P, 3P ≠ 0`.
+
+The equation is $$y^2 + (1-c)xy - by = x^3 - bx^2$$, and the point `P` is moved to `(0, 0)`.
+
+We also have the following formulas:
+* `P = (0, 0)`.
+* `2P = (b, bc)`.
+* `3P = (c, b-c)`.
+* `-P = (0, b)`.
+* `-2P = (b, 0)`.
+* `-3P = (c, c^2)`.
+
+Importantly, we have generalised the results here to work over any commutative ring `R`.
+
+## Main definitions and results
+
+* `IsTateNF W`: A typeclass saying that the Weierstrass curve `W` is in Tate normal form.
+* `toTateNF W`: Given a point `P` satisfying the condition `P, 2P, 3P ≠ 0`, this is a variable
+  change that brings the Weierstrass curve `W` to Tate normal form.
+
+## References
+
+* [James Parson, Moduli of Elliptic Curves](https://math.stanford.edu/~conrad/vigregroup/vigre03/moduli.pdf)
+-/
+
+noncomputable section
+
+namespace WeierstrassCurve
+
+/-- The Tate normal form is $$y^2 + (1-c)xy - by = x^3 - bx^2$$. -/
+@[mk_iff]
+class IsTateNF {R : Type*} [Zero R] (W : WeierstrassCurve R) : Prop where
+  a₂₃ : W.a₂ = W.a₃
+  a₄  : W.a₄ = 0
+  a₆  : W.a₆ = 0
+
+namespace Affine.Point
+
+variable {R : Type*} [CommRing R] {W : WeierstrassCurve R} (P : W.toAffine.Point)
+
+/-- Typeclass for a given point not being zero (the point at infinity). -/
+@[mk_iff]
+class NeZero : Prop where
+  neZero : P ≠ 0
```

Comment:
What's the problem which makes you not using system's [NeZero](https://leanprover-community.github.io/mathlib4_docs/Init/Data/NeZero.html#NeZero)?